atc/scheduler/factory/factory_parallel_test.go

```diff
@@ -0,0 +1,210 @@
+package factory_test
+
+import (
+	"github.com/concourse/concourse/atc"
+	"github.com/concourse/concourse/atc/scheduler/factory"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Factory Parallel", func() {
+	var (
+		buildFactory factory.BuildFactory
+
+		resources           atc.ResourceConfigs
+		resourceTypes       atc.VersionedResourceTypes
+		actualPlanFactory   atc.PlanFactory
+		expectedPlanFactory atc.PlanFactory
+	)
+
+	BeforeEach(func() {
+		actualPlanFactory = atc.NewPlanFactory(123)
+		expectedPlanFactory = atc.NewPlanFactory(123)
+
+		buildFactory = factory.NewBuildFactory(42, actualPlanFactory)
+
+		resources = atc.ResourceConfigs{
+			{
+				Name:   "some-resource",
+				Type:   "git",
+				Source: atc.Source{"uri": "git://some-resource"},
+			},
+		}
+
+		resourceTypes = atc.VersionedResourceTypes{
+			{
+				ResourceType: atc.ResourceType{
+					Name:   "some-custom-resource",
+					Type:   "registry-image",
+					Source: atc.Source{"some": "custom-source"},
+				},
+				Version: atc.Version{"some": "version"},
+			},
+		}
+	})
+
+	Context("when I have one parallel", func() {
+		It("returns the correct plan", func() {
+			actual, err := buildFactory.Create(atc.JobConfig{
+				Plan: atc.PlanSequence{
+					{
+						Parallel: &atc.PlanSequence{
+							{
+								Task: "some thing",
+							},
+							{
+								Task: "some other thing",
+							},
+						},
+						MaxInParallel: 1,
+					},
+				},
+			}, resources, resourceTypes, nil)
+			Expect(err).NotTo(HaveOccurred())
+
+			expected := expectedPlanFactory.NewPlan(atc.ParallelPlan{
+				MaxInParallel: 1,
+				Steps: []atc.Plan{
+					expectedPlanFactory.NewPlan(atc.TaskPlan{
+						Name:                   "some thing",
+						VersionedResourceTypes: resourceTypes,
+					}),
+					expectedPlanFactory.NewPlan(atc.TaskPlan{
+						Name:                   "some other thing",
+						VersionedResourceTypes: resourceTypes,
+					}),
+				},
+			})
+			Expect(actual).To(Equal(expected))
+		})
+	})
+
+	Context("when I have nested parallel steps", func() {
```

Comment:
I don't see how this can fail if the buildFactory can produce a plan from a `ParallelPlan`. I.e. the implementation is already identifying `Parallel` and recursing to produce a plan for substeps that may include another `parallel`. Delete?

Reply:
Agreed